### ISSUE TITLE
Random with shape.ordered causes global batch dim issue

### ISSUE THREAD

**Issue by Atticus1806** [rwth-i6/returnn_common#197]
I am currently getting an issue when using this module in my network : 
```python
class VariationalAutoEncoder(nn.Module):
  def __init__(
    self,
    out_size: int = 32,
    latent_size: int = 8,
  ):
    self.latent_dim = nn.FeatureDim("VAE_lat_dim", latent_size)
    self.lin_mu = nn.Linear(self.latent_dim)
    self.lin_log_var = nn.Linear(self.latent_dim)
    self.lin_out = nn.Linear(nn.FeatureDim("VAE_out_dim", out_size))

  def __call__(
    self, spectrogramm: nn.Tensor, spectrogramm_time: nn.SpatialDim
  ) -> [nn.Tensor, nn.Tensor, nn.Tensor]:
    x = nn.reduce(spectrogramm, mode="mean", axis=spectrogramm_time, use_time_mask=True)  # [B, F]

    # draw latent
    mu = self.lin_mu(x)  # [B, F]
    log_var = self.lin_log_var(x)
    std = nn.exp(0.5 * log_var)
    eps = nn.random_normal(std.shape_ordered)
    eps = eps * std
    eps = eps + mu
    out = self.lin_out(eps)
    return out, mu, log_var
```
Commenting / Skipping the nn.random_normal line gets the model working
Since the error is quite long I uploaded it to a gist: [Error](https://gist.github.com/Atticus1806/4398fafc9b39bd052b2b5bc95fd96cfc)

I am currently working on a simplified model but right now I am not  able to reproduce the error, maybe you already have an idea.
Maybe this again is related to the multiple calls of `get_network` during the use via NNetConstructor just as https://github.com/rwth-i6/returnn/issues/1069 ? 
Here is also the full model file: [Model](https://github.com/rwth-i6/i6_experiments/blob/main/users/hilmes/experiments/librispeech/nar_tts_2022/networks/tts_model.py)

**Comment by albertz** [comment]
Do you have the full log?

**Comment by albertz** [comment]
Do you have the generated config?

**Comment by albertz** [comment]
I reproduced the problem:
```python
def test_random_normal_shape_get_network():
  # https://github.com/rwth-i6/returnn_common/issues/197
  class _Net(nn.Module):
    def __init__(self):
      super().__init__()

    def __call__(self, x_: nn.Tensor) -> nn.Tensor:
      return x_ + nn.random_normal(x_.shape_ordered)

  from returnn.config import Config
  from returnn.tf.engine import Engine
  from returnn.datasets import init_dataset
  in_dim = nn.FeatureDim("in", 3)
  x = nn.Data("data", dim_tags=[nn.batch_dim, in_dim], available_for_inference=True)

  def _config_get_network(epoch: int, **_kwargs) -> dict:
    # noinspection PyStatementEffect
    epoch  # unused
    nn.reset_default_root_name_ctx()
    net = _Net()
    out = net(nn.get_extern_data(x))
    out.mark_as_default_output()
    out.mark_as_loss()
    net_dict = nn.get_returnn_config().get_net_dict_raw_dict(net)
    return net_dict

  config = Config({
    "task": "train", "num_epochs": 1, "start_epoch": 1,
    "get_network": _config_get_network,
    "extern_data": {x.name: {"dim_tags": [nn.batch_dim, in_dim], "available_for_inference": True}},
  })
  train_dataset = init_dataset(
    {"class": "DummyDataset", "input_dim": in_dim.dimension, "output_dim": 5, "num_seqs": 3})
  engine = Engine(config)
  engine.init_train_from_config(config, train_data=train_dataset)
  engine.train()
```
The test case is still a bit incomplete/wrong. The dataset here is wrong (it has time dim). Including a time dim causes a slightly different error, where I will need another additional fix.

**Comment by Atticus1806** [comment]
Here for completeness still the log and config:
https://gist.github.com/Atticus1806/a6ef2ebf966d481b361f06718fa1b3c9

**Comment by albertz** [comment]
I think this should be fixed now. Can you test?

**Comment by Atticus1806** [comment]
Yes works. Thanks for the quick fix!
